From: "Barak A. Pearlmutter" <barak+git@cs.nuim.ie>
Date: Wed, 27 Jul 2011 20:46:53 +0100
Subject: tweak Makefile

Use CFLAGS at link time.

Don't specify -O2 and -g (on by default) so they can be overwritten
by builder, e.g., dpkg-buildflags.

Use CPPFLAGS.  Needed during compilation to call hardened routines
during hardened build.  See lintian tag
hardening-no-fortify-functions.

Allow parallelism during "make install" by ensuring that each
installation stanza creates any directories it targets.  This might
result in multiple creation; that is okay.

Use gzip -n option for reproducible builds.

Install s9core.a library in multiarch directory as libs9core.a
---
 Makefile | 44 ++++++++++++++++++++++++++++++--------------
 1 file changed, 30 insertions(+), 14 deletions(-)

diff --git a/Makefile b/Makefile
index 01e74cc..58b2c82 100644
--- a/Makefile
+++ b/Makefile
@@ -3,8 +3,8 @@
 # By Nils M Holm, 2007-2018
 # In the public domain
 
-# Change at least this line:
-PREFIX= /u
+prefix=/usr/local
+PREFIX=$(prefix)
 
 # Base version and Release
 BASE=		20181030
@@ -45,31 +45,40 @@ EXTRA_LIBS+=
 #				# (requires the Curses extension)
 
 DEFS=	$(OSDEF) \
-	-DLIBRARY_PATH="\".:~/s9fes:$(S9DIR)\"" \
-	-DIMAGE_DIR="\"$(S9DIR)\"" \
+	$(CPPFLAGS) \
+	-DLIBRARY_PATH="\".:~/s9fes:$(pkgdatadir)\"" \
+	-DIMAGE_DIR="\"$(pkgdatadir)\"" \
 	-DEXTENSIONS="$(EXTRA_INIT)" \
 	-DNETWORK \
 	-DCURSES_COLOR \
 	-DCURSES_RESET
 
 # Where to install the stuff
-S9DIR=	$(PREFIX)/share/s9fes
-BINDIR=	$(PREFIX)/bin
-INCDIR=	$(PREFIX)/include
-LIBDIR=	$(PREFIX)/lib
-MANDIR=	$(PREFIX)/man/man1
+PACKAGE=scheme9
+exec_prefix=$(prefix)
+bindir=$(exec_prefix)/bin
+includedir=$(prefix)/include
+datadir=$(prefix)/share
+pkgdatadir=$(datadir)/$(PACKAGE)
+libdir=$(exec_prefix)/lib/$(shell dpkg-architecture -q DEB_HOST_MULTIARCH || echo .)
+mandir=$(prefix)/man
+S9DIR=$(DESTDIR)$(pkgdatadir)
+BINDIR=$(DESTDIR)$(bindir)
+INCDIR=$(DESTDIR)$(includedir)
+LIBDIR=$(DESTDIR)$(libdir)
+MANDIR=$(DESTDIR)$(mandir)/man1
 
 # Set up environment to be used during the build process
 BUILD_ENV=	env S9FES_LIBRARY_PATH=.:lib:ext/sys-unix:ext/curses:ext/csv:contrib S9FES_IMAGE_DIR=.
 
 SETPREFIX=	sed -e "s|^\#! /usr/local|\#! $(PREFIX)|"
 
-default:	s9 s9.image s9.1.gz s9.1.txt libs9core.a
+default:	s9 s9.image s9.1 libs9core.a
 
 all:	default
 
 s9:	s9.o s9core.o $(EXTRA_OBJS)
-	$(CC) -o s9 $(LDFLAGS) s9.o s9core.o $(EXTRA_OBJS) $(EXTRA_LIBS)
+	$(CC) -o s9 $(CFLAGS) $(LDFLAGS) s9.o s9core.o $(EXTRA_OBJS) $(EXTRA_LIBS)
 
 s9.o:	s9.c s9core.h s9import.h s9ext.h
 	$(CC) -o s9.o $(CFLAGS) $(DEFS) -c s9.c
@@ -84,8 +93,13 @@ s9.image:	s9 s9.scm ext/sys-unix/unix.scm ext/curses/curses.scm \
 libs9core.a: s9core.o
 	ar q libs9core.a s9core.o
 
-s9.1.gz:	s9.1
-	sed -e "s,@S9DIR@,$(S9DIR)," <s9.1 |gzip -9 >s9.1.gz
+%.1: %.1.in
+	sed \
+	 -e "s,@S9DIR@,$(pkgdatadir),g" \
+	 < $@.in > $@
+
+%.gz: %
+	gzip -9 -n < $* > $@
 
 unix.o:	ext/sys-unix/unix.c s9core.h s9import.h s9ext.h
 	$(CC) $(CFLAGS) $(DEFS) -I . -o unix.o -c ext/sys-unix/unix.c
@@ -136,7 +150,7 @@ install-s9:	s9 s9.scm s9.image s9.1.gz libs9core.a
 	install -d -m 0755 $(INCDIR)
 	install -d -m 0755 $(MANDIR)
 	install $C -m 0755 s9 $(BINDIR)
-	strip $(BINDIR)/s9
+	echo strip $(BINDIR)/s9
 	install $C -m 0644 s9.scm $(S9DIR)
 	install $C -m 0644 s9.image $(S9DIR)
 	install $C -m 0644 lib/* $(S9DIR)
@@ -151,6 +165,7 @@ install-s9:	s9 s9.scm s9.image s9.1.gz libs9core.a
 	install $C -m 0644 s9import.h $(INCDIR)
 
 install-util:
+	install -d -m 0755 $(BINDIR)
 	$(SETPREFIX) <prog/s9help.scm >$(BINDIR)/s9help
 	$(SETPREFIX) <prog/s9resolve.scm >$(BINDIR)/s9resolve
 	$(SETPREFIX) <prog/scm2html1.scm >$(BINDIR)/scm2html
@@ -161,6 +176,7 @@ install-util:
 		  $(BINDIR)/scmpp
 
 install-progs:
+	install -d -m 0755 $(BINDIR)
 	$(SETPREFIX) <prog/advgen.scm >$(BINDIR)/advgen
 	$(SETPREFIX) <prog/c2html1.scm >$(BINDIR)/c2html
 	$(SETPREFIX) <prog/cols.scm >$(BINDIR)/cols
